azure/services/managedclusters/spec.go

```diff
@@ -384,13 +380,50 @@ func (s *ManagedClusterSpec) getManagedClusterVersion(existing *asocontainerserv
 	return versions.GetHigherK8sVersion(s.Version, *existing.Status.CurrentKubernetesVersion)
 }
 
+// ResourceRef implements azure.ASOResourceSpecGetter.
+func (s *ManagedClusterSpec) ResourceRef() genruntime.MetaObject {
+	if s.Preview {
+		return &asocontainerservicev1preview.ManagedCluster{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: s.Name,
+			},
+		}
+	}
+	return &asocontainerservicev1.ManagedCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: s.Name,
+		},
+	}
+}
+
 // Parameters returns the parameters for the managed clusters.
 //
 //nolint:gocyclo // Function requires a lot of nil checks that raise complexity.
-func (s *ManagedClusterSpec) Parameters(ctx context.Context, existing *asocontainerservicev1.ManagedCluster) (params *asocontainerservicev1.ManagedCluster, err error) {
+func (s *ManagedClusterSpec) Parameters(ctx context.Context, existingObj genruntime.MetaObject) (params genruntime.MetaObject, err error) {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "managedclusters.Service.Parameters")
 	defer done()
 
+	// If existing is preview, convert to stable then back to preview at the end of the function.
+	var existing *asocontainerservicev1.ManagedCluster
+	var existingStatus asocontainerservicev1preview.ManagedCluster_STATUS
+	if existingObj != nil {
+		if s.Preview {
+			existingPreview := existingObj.(*asocontainerservicev1preview.ManagedCluster)
+			existingStatus = existingPreview.Status
+			hub := &asocontainerservicev1hub.ManagedCluster{}
+			if err := existingPreview.ConvertTo(hub); err != nil {
+				return nil, err
+			}
+			stable := &asocontainerservicev1.ManagedCluster{}
+			if err := stable.ConvertFrom(hub); err != nil {
+				return nil, err
+			}
+			existing = stable.DeepCopy()
+		} else {
+			existing = existingObj.(*asocontainerservicev1.ManagedCluster)
+		}
+	}
+
```

Comment:
@nojnhuh it appears that the `// Additional tags managed separately` comment refers to functionality that this PR has broken (until we fix it, that is!)

Reply:
This should be fixed by this commit: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4617/commits/5ff390c196e19380a40af77971d88e58ab6f66d0#diff-d7496d9845d55dda5142aab2970a5a9aa1012fbb52d2c462cc5b220682ca520aR821

The issue was caused by not updating the type of the `TagsGetterSetter` to `genruntime.MetaObject`, and handling tags differently based on the api version.